src/crystal/system/win32/hostname.cr

```diff
@@ -0,0 +1,12 @@
+require "c/sysinfoapi"
+
+module Crystal::System
+  def self.hostname
+    name_size = LibC::MAX_COMPUTER_NAME_SIZE.to_u32
+    unless LibC.GetComputerNameExW(LibC::COMPUTER_NAME_FORMAT::ComputerNameDnsHostname, out machine_name, pointerof(name_size))
```

Comment:
According to the description of [GetComputerNameExW](https://docs.microsoft.com/en-us/windows/desktop/api/sysinfoapi/nf-sysinfoapi-getcomputernameexw), I'm not sure this usage is correct. Might be reading it wrong... But `MAX_COMPUTER_NAME_SIZE` doesn't make sense as an size limit, since you're not providing a buffer. And the name could actually have more characters because it's a DNS name.

Reply:
According to the docs, yes:
- it needs the buffer to be allocated —`out` means that the function will allocate it which it doesn't;
- it could be called twice, once with `lpBuffer` as `nil` so that it will set `nSize`, so we can allocate a properly sized buffer and call the function again that will fill the value.

Reply:
It's a bit weird of API design on MS's part, imo. I'd prefer if the function just returned name length (and perhaps -1 if the call failed.) 

@ysbaddaden is right. 

DNS names have a 255 character limit [IIRC](https://blogs.msdn.microsoft.com/oldnewthing/20120412-00/?p=7873), so I just initialized it to the max since the function wants it allocated beforehand and will set the value accordingly. Since I use a StaticArray on the other side, technically I'm wasting a tiny bit of memory (since the DNS name will probably be much less than 255 chars) but I do think that this is a bit cleaner than 2 calls.

Reply:
It's a standard pattern for windows, you can use [`retry_wstr_buffer`](https://github.com/crystal-lang/crystal/blob/master/src/crystal/system/windows.cr#L3) for this.

Reply:
Would this be the correct way to use `retry_wstr_buffer` since `GetComputerNameExW` wants a pointer to the buffer length? This compiles and runs.
```
def hostname
  retry_wstr_buffer do |buffer, small_buf|
    name_size = buffer.size.to_u
    if LibC.GetComputerNameExW(LibC::COMPUTER_NAME_FORMAT::ComputerNameDnsHostname, buffer, pointerof(name_size))
      return String.from_utf16(buffer[0, name_size])
    else
      raise WinError.new("Failed to get machine hostname")
    end
  end
end
```

Reply:
No, you need to either `break` the final `String`, `next` with the next size of the buffer, or raise. You have no provision for actually retrying, making `retry_wstr_buffer` pointless.

Reply:
I was initially a little confused by what the two`yield` statements were doing. I know that they yield execution, but one was used for size and the other for the content of the buffer. I think I understand now, thanks for explaining. Newest PR has these changes.